### PR TITLE
embeddings: Updated RPC function and English detection mechanism

### PR DESCRIPTION
**Pull request by vaiibhavgupta**

### CODE REVIEW

File: src/embedding.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        upsert_response = db.table("embeddings").upsert({
+            "text": input_body['text'],
+            "module": input_body['module'],
+            "chapter": input_body['chapter'],
+            "page": input_body['page'],
+            "chunk": input_body['chunk'],
+            "content": input_body['content'],
+            "embedding": embedding
+        }).execute()
+        return {'content' : {'message' : upsert_response.data}, 'status_code': 201}
```
Comment (langdonholmes) [inline src/embedding.py:22]: Please use FastAPI for this: https://fastapi.tiangolo.com/reference/response/

This is for readability and for consistency across the codebase.

File: models/embedding.py [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
+
+class RetrievalInput(InputBase):
+    text: Optional[str]
+    page: Optional[str]
+    content: str # text to compare to (student summary)
 
+class Match(BaseModel):
+    chunk_slug: str
+    content: str
+    similarity: float
```
Comment (langdonholmes) [inline models/embedding.py:21]: Please add optional params (with defaults) here for `similarity_threshold` and `match_count`.

File: src/embedding.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    except Exception as ex:
+        return {'content' : {'message' : ex.message}, 'status_code': ex.code}
+
+async def retrieve_chunks(input_body: RetrievalInput, match_threshold: float = 0.3, match_count: int = 2) -> RetrievalResults:
+    content_embedding = embedding_pipeline(input_body['content'])
+    try:
```
Comment (langdonholmes) [inline src/embedding.py:28]: Please assemble query params outsides of the try/except block.

File: src/main.py [hunk 4]
```diff
@@ -68,7 +68,7 @@ async def score_answer(input_body: AnswerInput) -> AnswerResults:
 
 
 @app.post("/generate/embedding")
-async def gen_embedding(input_body: ChunkInput) -> ChunkEmbedding:
+async def gen_embedding(input_body: ChunkInput) -> dict[str, Any]:
```
Comment (langdonholmes) [inline src/main.py:71]: The return type should be a `fastapi.Response`.

File: src/summary_eval.py [hunk 5]
```diff
@@ -144,8 +146,13 @@ async def summary_score(summary_input: SummaryInput) -> SummaryResults:
     summary.score_similarity()
     summary.suggest_keyphrases()
 
+    summary.results["english"] = True
+    is_detection_reliable, _, details = cld2.detect(summary_input.summary)
+    if not is_detection_reliable or details[0][0] != "ENGLISH":
```
Comment (langdonholmes) [inline src/summary_eval.py:151]: I think there is a mistake in the logic here. We want to return English unless we have a reliable non-English result.

A minimal change to produce the desired result is changing the if condition to:
`if is_detection_reliable and details[0][0] != "ENGLISH":`

File: src/summary_eval_supabase.py [hunk 6]
```diff
@@ -162,8 +163,13 @@ async def summary_score_supabase(summary_input: SummaryInput) -> SummaryResults:
     summary.score_similarity()
     summary.suggest_keyphrases()
 
+    summary.results["english"] = True
+    is_detection_reliable, _, details = cld2.detect(summary_input.summary)
+    if not is_detection_reliable or details[0][0] != "ENGLISH":
```
Comment (langdonholmes) [inline src/summary_eval_supabase.py:168]: Thanks for updating both files. Please see my other comment about this if statement.

File: src/summary_eval.py (hunk 5 above)
Comment (vaiibhavgupta) [reply src/summary_eval.py:151]: @langdonholmes My logic was to mark language as not English if `detection is not reliable (no matter the language detected)` OR `language detected is not English`. But, going by your logic, it seems we want to assume language as English and want to mark a summary as junk only when `detection is reliable` AND `language is not english`. I have changed the logic.

**Comment by vaiibhavgupta** [comment]
@langdonholmes I just updated test cases for `generate_embedding` and `retrieve_chunks`. Can you please review them?

FYI, I am unable to test them in my local by running `python test_main.py`. I am having syntax issues like `models not found` (the `models` folder. However, I am able to test the two test functions from within a jupyter notebook and they are running fine.

**Comment by langdonholmes** [comment]
@vaiibhavgupta 

This is pretty close. I have some work to do before we deploy, so I can take it from here. The following comments are for your future reference. I will make the changes myself.

1. In the future, please note that testing with Jupyter notebooks is not sufficient. We need to get this sorted if you are going to continue contributing to this repository. Please send me details of the errors you ran into (on Slack). This includes the command you ran, the output, and any troubleshooting steps you took.
2. We are still not using `similarity_threshold` or `match_count`. A fringe benefit of adding this feature in is that we can write a test with an arbitrarily high `similarity_threshold` (1.0) to test that everything works when no matches are returned.
3. Your try/except blocks are problematic for two reasons: 1) please avoid catching all exceptions with `Exception`. It is almost always better to specify which exception you are expecting to catch. 2) The try block includes two things that can fail: the database query and indexing the (possibly empty) result. These should be handled separately.

Thanks for your help with this!

**Pull request merged**